### DISCUSSION TITLE
2023-10-23 백엔드 정기 회의

### DISCUSSION THREAD

**Discussion by Jake1152** [jiphyeonjeon-42/backend#787] [General]
6차 개발 사항 논의, 개발 사항 분담

- v2에 리팩터링 진행
  - Auth api 작성 및 DB이관
  - typeORM를 Kysely 쿼리 빌더로 변경 
  - Reviews 테스트 로직 추가
  - 타입안정성 높이기
  - front에서도 따로 정의 필요없이 backend에서 정의한 것을 가져다 쓰는 식으로 변경
    - typescript type 사용
    - contract를 패키지로 만들어서 사용 가능

**Comment by Jake1152** [comment]
- Auth 재작성 및 DB table 이관

**Reply by Jake1152** [reply]
# 담당
- jimin, jim

# 해야할 일
- 구글 Auth로 기존 사이트에서 비밀번호 입력받아서 회원가입 할 수 있게 하던 부분을 대체함

**Reply by Jake1152** [reply]
- 다른 계정임에도 'jwoo' 아이디로 로그인 되는 이슈
- 사서 권한이 뺴앗겨지는 경우
  - typeORM, findOne()함수에서 첫번쨰 유저를 가져와서 생기는 문제로 추정 
![image](https://github.com/jiphyeonjeon-42/backend/assets/74135462/2b9128d9-daf1-469e-97f2-04de9f9a86ff)

**Comment by Jake1152** [comment]
typeORM를 Kysely 쿼리 빌더로 변경

**Reply by Jake1152** [reply]
왜 typeORM에서 바꾸는가?
1.  런타임에서의 타입을 지정
- 현재는 any 로 나올 때가 있어서 as를  써서 타입을 바꾸어야 했었다
- 런타임에서도 타입을 맞추어준다.

2. 객체 속 프로퍼티가 undeifined로 나오는 이슈
- User라는 객체 안에 일부 프로퍼티가 undefined로 나온다.

3. ORM에서 불필요하게 처리해주는 것들이 있다.
- 내부적으로connection control, caching, 객체 함수로 되어있는 것
- typeORM 내부적으로 많은 동작을 해주는데 어디까지 동작해주는지 모르며, 현재 우리 규모에서는 필요없다 생각
- e.g)typeORM User 객체 내부에 객체를 구성하기 위한 수많은 함수들 같은 것

**Reply by Jake1152** [reply]
- 다시 확인해보면 좋겠다 생각드는 것
  - typeORM에서 어떤 케이스에 any로 나오게 되는가
  - typeORM의 문제인지 혹은 다른 것 때문에 발생하는 문제인지 
  - 매개변수 인자가 불안정한지

**Reply by Jake1152** [reply]
@jimin52 @nyj001012 @JeongJiHwan @scarf005   typeORM에서 any로 나오게 되는 경우 상황이 정확히 어떻게 되나요?
다른분이랑 이야기하다가 들게된 의문인데 typeORM이 type을 위해 있고,  typeORM이 node베이스의 백앤드를 쓰는 기업들에서도 ORM으로 잘 쓰이는 것으로 보였습니다

typeORM에서 any로 결과가 나오는 경우가 혹시 저희가 놓치는 부분은 있을 수도 있지 않을까하여 어떤 경우에 any로 나오게되는지 질문드리게 되었습니다

**Reply by Jake1152** [reply]
예나님의 답변:

user table 이 있다고 했을 때
닉네임, 비밀번호 등 일부 컬럼만 조회할때 any로 나온다.
join에서도 마찬가지로 정의한 것 이외에도 any로 나오게된다

타입 안정성 보장을 위해 join한 결과를 타입 지정함으로써 as로 만들어서 타입안정성을 만들 수도 있다.
하지만 코딩의 품이 들게되는 단점이 있다.
kysely를 쓰면 entity를 지정하는 것처럼 컬럼 타입 지정 가능한다
자동으로 알아서 컬럼에 맞는 타임을 지정해준다.

**Comment by Jake1152** [comment]
Reviews 테스트 로직 추가

**Reply by Jake1152** [reply]
다른 곳에서도 테스트 코드 추가해야하지만 
우선은 Reviews 테스트 로직 추가부터 진행함

**Reply by Jake1152** [reply]
- jest
- node test runner

**Reply by Jake1152** [reply]
- jest 쓰려던 이유 
 test runner의 존재 유무를 몰랐으며 package.json에 jest가 있어서 쓰려고 함
  
하지만 test runner에서 mocking도 되고 하려는 것들이 대체로 되는 것으로 보이므로 
테스트는 test runner로 결정됨

**Reply by Jake1152** [reply]
- jest와 node test runner의 구체적인 차이점, 실무에서는 어떤 것을 더 많이 쓰고 그 이유는 무엇인가?

**Comment by Jake1152** [comment]
타입안정성 높이기

**Reply by Jake1152** [reply]
zod 로 대체, 기존코드에 any로 된 부분들이 많았다. 이런 부분들을 다 타입을 지정해서 처리

**Comment by Jake1152** [comment]
front에서도 따로 정의 필요없이 backend에서 정의한 것을 가져다 쓰는 식으로 변경

**Reply by Jake1152** [reply]
frontend,backend 둘이 공통으로 쓰는 type을 contract로 패키지화하여 처리

**Comment by Jake1152** [comment]
# TO DO
- 현재 개발에서 필요한 기술 스택 알아보고 실습
  - kysely, zod, contract, node test runner
- 10.30월요일 전까지!

**Comment by scarf005** [comment]
정리 자세하게 해주셔서 감사합니다! 많이 도움 드리지 못해 죄송합니다...

**Reply by Jake1152** [reply]
아닙니다 ㅎ 질문해주신덕에 test runner알게 되었어여 ㅋㅋ
